subxt/src/extrinsic/extra.rs

```diff
@@ -1,481 +0,0 @@
-    ) -> Self;
-
-    /// Returns the transaction extra.
-    fn extra(&self) -> Self::Extra;
-}
-
-/// Default `SignedExtra` for substrate runtimes.
-#[derive(Derivative, Encode, Decode, TypeInfo)]
-#[derivative(
-    Clone(bound = ""),
-    PartialEq(bound = ""),
-    Debug(bound = ""),
-    Eq(bound = "")
-)]
-#[scale_info(skip_type_params(T))]
-pub struct DefaultExtraWithTxPayment<T: Config, X> {
-    spec_version: u32,
-    tx_version: u32,
-    nonce: T::Index,
-    genesis_hash: T::Hash,
-    marker: PhantomDataSendSync<X>,
-}
-
-impl<T, X> SignedExtra<T> for DefaultExtraWithTxPayment<T, X>
-where
-    T: Config,
-    X: SignedExtension<AccountId = T::AccountId, Call = ()> + Default,
-{
-    type Extra = (
-        CheckSpecVersion<T>,
-        CheckTxVersion<T>,
-        CheckGenesis<T>,
-        CheckMortality<T>,
-        CheckNonce<T>,
-        CheckWeight<T>,
-        X,
-    );
-    type Parameters = ();
-
-    fn new(
-        spec_version: u32,
-        tx_version: u32,
-        nonce: T::Index,
-        genesis_hash: T::Hash,
-        _params: Self::Parameters,
-    ) -> Self {
-        DefaultExtraWithTxPayment {
-            spec_version,
-            tx_version,
-            nonce,
-            genesis_hash,
-            marker: PhantomDataSendSync::new(),
-        }
-    }
-
-    fn extra(&self) -> Self::Extra {
-        (
-            CheckSpecVersion(PhantomDataSendSync::new(), self.spec_version),
-            CheckTxVersion(PhantomDataSendSync::new(), self.tx_version),
-            CheckGenesis(PhantomDataSendSync::new(), self.genesis_hash),
-            CheckMortality(
-                (Era::Immortal, PhantomDataSendSync::new()),
-                self.genesis_hash,
-            ),
-            CheckNonce(self.nonce),
-            CheckWeight(PhantomDataSendSync::new()),
-            X::default(),
-        )
-    }
-}
-
-impl<T, X: SignedExtension<AccountId = T::AccountId, Call = ()> + Default> SignedExtension
-    for DefaultExtraWithTxPayment<T, X>
-where
-    T: Config,
-    X: SignedExtension,
-{
-    const IDENTIFIER: &'static str = "DefaultExtra";
-    type AccountId = T::AccountId;
-    type Call = ();
-    type AdditionalSigned =
-        <<Self as SignedExtra<T>>::Extra as SignedExtension>::AdditionalSigned;
-    type Pre = ();
-
-    fn additional_signed(
-        &self,
-    ) -> Result<Self::AdditionalSigned, TransactionValidityError> {
-        self.extra().additional_signed()
-    }
-    fn pre_dispatch(
-        self,
-        _who: &Self::AccountId,
-        _call: &Self::Call,
-        _info: &DispatchInfoOf<Self::Call>,
-        _len: usize,
-    ) -> Result<Self::Pre, TransactionValidityError> {
-        Ok(())
-    }
-}
-
-/// A default `SignedExtra` configuration, with [`ChargeTransactionPayment`] for tipping.
-///
-/// Note that this must match the `SignedExtra` type in the target runtime's extrinsic definition.
-pub type DefaultExtra<T> = DefaultExtraWithTxPayment<T, ChargeTransactionPayment<T>>;
```

Comment:
I am wondering how this single `DefaultExtra` seems to work for both polkadot (in the examples) and substrate (in the tests), because the former just expects a `Compact<u128>` for the tip section in the extra params, and the latter expects a `(Compact<u128>, Option<u32>)`, which is what this provides via `ChargeTransactionPayment`. Am I missing something or did we get lucky somehow with that extra `0` byte for the `Option<u32> = None` being put into the transaction extras in our examples?

Reply:
`ChargeTransactionPayment` comes last, so I guess that the node will successfully decode the `Compact<u128>` and ignore the subsequent data. This data isn't included in `AdditionalSigned` so won't affect the signature.

Reply:
I'd have thought that decoding the call data which comes next would be naffed up in that case! It's a mystery to me at the moment; I'd have to experiment a bit!

Reply:
(Also, the "extra" data is a part of the payload that is signed, so the signature would be different too?)

Reply:
Not sure what's going on here then...